### DISCUSSION TITLE
Delete related entry conditionally with the actual object

### DISCUSSION THREAD

**Discussion by SarenT** [prisma/prisma#22649] [Q&A] [answered]
### Question

Hi,
I have an author table and a post table. I want the user to decide, if they want to delete the related posts, when they are deleting the author entry. So I want to conditionally delete posts.

```
this.prisma.author.delete({
                where: {
                  id: id,
                },
                data: {posts: {deleteMany: {}}}
              });
```
This is giving me an error that data is expected to be "never". I guess, `delete` does not expect such a field. How can I ensure this with a transactional query? I wanted to use nested query to make sure either this fails as a whole or succeeds as a whole.
`update` works as documented. But I do not want to update `author`, I want to delete it.
I can as well make this operation conditional with a boolean value:
```
{...(delete_posts ? 
                    {posts: {deleteMany: {}}}: 
                    {})}
```
But I was not able to delete them together.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @SarenT 👋 

Thank you for raising this question.

The [delete](https://www.prisma.io/docs/orm/reference/prisma-client-reference#delete) operation does not support passing `data` as part of the options. To perform a conditional delete of an author and their related posts within a transaction, you can use [`$transaction`](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#the-transaction-api)API. This allows you to run multiple queries together, ensuring that either all of them succeed or none of them do.

You can delete an `author` and all their `posts` with two separate queries in a transaction. Here's how you can structure your transaction to conditionally delete the posts and then the author:

```js
const prisma = new PrismaClient();

async function deleteAuthorAndPosts(authorId, deletePosts) {
  const queries = [];

  if (deletePosts) {
    // Add the deleteMany query for posts to the transaction if posts should be deleted
    queries.push(prisma.post.deleteMany({
      where: { authorId: authorId },
    }));
  }

  // Add the delete query for the author to the transaction
  queries.push(prisma.author.delete({
    where: { id: authorId },
  }));

  // Execute the transaction
  try {
    const transaction = await prisma.$transaction(queries);
    console.log('Transaction succeeded:', transaction);
  } catch (error) {
    console.error('Transaction failed:', error);
  }
}

// Usage example
deleteAuthorAndPosts(7, true); // Replace 7 with the actual author ID and true with the condition to delete posts
```

The transaction will ensure that both operations (deleting posts and deleting the author) are treated as a single unit of work.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by SarenT** [reply]
Thanks a lot for the detailed answer and sample code!
